server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java

```diff
@@ -68,34 +82,109 @@ public Session getSession(DataReadRequest request) {
-                    .datasetVersion(String.valueOf(request.getDatasetVersionId()))
-                    .tableName(request.getTableName())
-                    .start(request.getStart())
-                    .startInclusive(request.isStartInclusive())
-                    .end(request.getEnd())
-                    .endInclusive(request.isEndInclusive())
-                    .batchSize(request.getBatchSize())
-                    .build();
+                .sessionId(request.getSessionId())
+                .datasetName(request.getDatasetName())
+                .datasetVersion(String.valueOf(request.getDatasetVersionId()))
+                .tableName(request.getTableName())
+                .start(request.getStart())
+                .startInclusive(request.isStartInclusive())
+                .end(request.getEnd())
+                .endInclusive(request.isEndInclusive())
+                .batchSize(request.getBatchSize())
+                .build();
         // insert session
         sessionDao.insert(session);
-        // get data index
-        List<DataIndex> dataIndices = dataIndexProvider.returnDataIndex(
-                QueryDataIndexRequest.builder()
-                    .tableName(session.getTableName())
-                    .batchSize(session.getBatchSize())
-                    .start(session.getStart())
-                    .startInclusive(session.isStartInclusive())
-                    .end(session.getEnd())
-                    .endInclusive(session.isEndInclusive())
-                    .build()
-        );
-        Long sid = session.getId();
+        return session;
+    }
+
+    @Async
+    public void generateDataReadLog(Long sessionId) {
+        try {
+            this.generate(sessionId);
+        } catch (Exception e) {
+            log.error("Error while generate data read logs for session: {}", sessionId, e);
+            failSessionQueue.add(new FailSession(sessionId));
+        }
+    }
+
+    private static class FailSession implements Delayed {
+        int failCount = 0;
+
+        Long sessionId;
+
+        FailSession(Long sessionId) {
+            this.sessionId = sessionId;
+        }
+
+        public long getDelayTime() {
+            long delay = 100;
+            if (failCount > 0) {
+                delay *= (2L << (failCount - 1));
+            }
+            return delay;
+        }
+
+        @Override
+        public long getDelay(@NotNull TimeUnit unit) {
+            return unit.convert(getDelayTime(), TimeUnit.MILLISECONDS);
+        }
+
+        @Override
+        public int compareTo(@NotNull Delayed o) {
+            long diffMillis = getDelay(TimeUnit.MILLISECONDS) - o.getDelay(TimeUnit.MILLISECONDS);
+            diffMillis = Math.min(diffMillis, 1);
+            diffMillis = Math.max(diffMillis, -1);
+            return (int) diffMillis;
+        }
+    }
+
+    @Scheduled(fixedRate = 1000)
+    public void dealWithErrorSessions() {
+        while (!failSessionQueue.isEmpty()) {
```

Comment:
Dead loop when it is not empty and no element with expired delay?

Reply:
fixed